[JSC] Move Promise.any to C++

Constellation opened the pull request:
#### 7c3a43aeff3abf9c456b20ca8514c329013c3dcb
<pre>
[JSC] Move Promise.any to C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=303732">https://bugs.webkit.org/show_bug.cgi?id=303732</a>
<a href="https://rdar.apple.com/166035878">rdar://166035878</a>

Reviewed by Sosuke Suzuki.

This patch moves Promise.any implementation from JS to C++ in the same
way to Promise.all and Promise.allSettled. We use existing helpers and
making it much faster than old Promise.any implementation. We define
internal microtask for major case, so no longer need to invoke a custom
function through Interpreter. Also we rename JSPromiseAllContext to
JSPromiseCombinatorsContext since it is used by Promise.all /
Promise.allSettled / Promise.any at this point.

                               ToT                     Patched

    promise-any          10.2568+-0.3150     ^      6.4039+-0.0927        ^ definitely 1.6016x faster

Test: JSTests/microbenchmarks/promise-any.js

* JSTests/microbenchmarks/promise-any.js: Added.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(any): Deleted.
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsContext.cpp: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllContext.cpp.
(JSC::JSPromiseCombinatorsContext::create):
(JSC::JSPromiseCombinatorsContext::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsContext.h: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllContext.h.
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsContextInlines.h: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllContextInlines.h.
(JSC::JSPromiseCombinatorsContext::createStructure):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsGlobalContext.cpp: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.cpp.
(JSC::JSPromiseCombinatorsGlobalContext::create):
(JSC::JSPromiseCombinatorsGlobalContext::createStructure):
(JSC::JSPromiseCombinatorsGlobalContext::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsGlobalContext.h: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.h.
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::promiseAllSlow):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::promiseAllSettledSlow):
(JSC::promiseAnySlow):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.h:
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::promiseAnyRejectFunctionExecutableSlow):
(JSC::VM::promiseAnySlowRejectFunctionExecutableSlow):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseAnyRejectFunctionExecutable):
(JSC::VM::promiseAnySlowRejectFunctionExecutable):

Canonical link: <a href="https://commits.webkit.org/304084@main">https://commits.webkit.org/304084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9284557d4a2131521da17cba58e35426234fe0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134547 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7005 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45779 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142073 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86507 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0ae7a2bc-4266-4235-a32b-9806fa9e41c9) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7608 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6865 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e1442060-df3f-4a3d-b372-9fd719aec44b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137494 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5320 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120588 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/83643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fb9fad37-9b0e-41b2-ba56-f372e58c3116) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5179 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2796 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo-libwebrtc~~](https://ews-build.webkit.org/#/builders/166/builds/2680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126595 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144767 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133055 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6681 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39311 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6755 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5605 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111515 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5011 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116863 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60539 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20769 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6732 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35058 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165973 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6547 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70314 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43384 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6781 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6658 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSMicrotask.cpp:
```diff
@@ -378,6 +379,45 @@ void runInternalMicrotask(JSGlobalObject* globalObject, InternalMicrotask task,
+    case InternalMicrotask::PromiseAnyResolveJob: {
+        auto* promise = jsCast<JSPromise*>(arguments[0]);
+        JSValue resolution = arguments[1];
+        auto* context = jsCast<JSPromiseAllContext*>(arguments[3]);
```
`JSPromiseAllContext` is no longer for only `Promise.all` so should we rename it to kind of `JSPromiseCombinatorsContext` or any other better name?

Constellation replied:
Sounds good. Changed.

sosukesuzuki reviewed (approved):
r=me, with comments

Constellation committed:
[JSC] Move Promise.any to C++
https://bugs.webkit.org/show_bug.cgi?id=303732
rdar://166035878

Reviewed by Sosuke Suzuki.

This patch moves Promise.any implementation from JS to C++ in the same
way to Promise.all and Promise.allSettled. We use existing helpers and
making it much faster than old Promise.any implementation. We define
internal microtask for major case, so no longer need to invoke a custom
function through Interpreter. Also we rename JSPromiseAllContext to
JSPromiseCombinatorsContext since it is used by Promise.all /
Promise.allSettled / Promise.any at this point.

                               ToT                     Patched

    promise-any          10.2568+-0.3150     ^      6.4039+-0.0927        ^ definitely 1.6016x faster

Test: JSTests/microbenchmarks/promise-any.js

* JSTests/microbenchmarks/promise-any.js: Added.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(any): Deleted.
* Source/JavaScriptCore/heap/Heap.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSFunctionWithFields.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsContext.cpp: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllContext.cpp.
(JSC::JSPromiseCombinatorsContext::create):
(JSC::JSPromiseCombinatorsContext::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsContext.h: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllContext.h.
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsContextInlines.h: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllContextInlines.h.
(JSC::JSPromiseCombinatorsContext::createStructure):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsGlobalContext.cpp: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.cpp.
(JSC::JSPromiseCombinatorsGlobalContext::create):
(JSC::JSPromiseCombinatorsGlobalContext::createStructure):
(JSC::JSPromiseCombinatorsGlobalContext::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSPromiseCombinatorsGlobalContext.h: Renamed from Source/JavaScriptCore/runtime/JSPromiseAllGlobalContext.h.
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::promiseAllSlow):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::promiseAllSettledSlow):
(JSC::promiseAnySlow):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.h:
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/Microtask.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::promiseAnyRejectFunctionExecutableSlow):
(JSC::VM::promiseAnySlowRejectFunctionExecutableSlow):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::promiseAnyRejectFunctionExecutable):
(JSC::VM::promiseAnySlowRejectFunctionExecutable):

Canonical link: https://commits.webkit.org/304084@main
